### PR TITLE
Skip to frame

### PR DESCRIPTION
**Pull request by davidje13**
Add capability to skip to an entered frame (including rewinding)

Extracted from #22

### CODE REVIEW

File: engine/GameOrchestrator.js [hunk 1]
```diff
@@ -109,11 +112,50 @@ define([
...
 						type,
 						steps,
+						checkbackTime: this.config.play.checkbackTime,
 					});
 				}, this.gameActive);
 			}
 		}
 
+		skip() {
+			if (this.config.game.skipFrame === undefined) {
```
Comment (davidje13) [inline engine/GameOrchestrator.js]: The code in this function seems to be 99% duplicated from `step()`. Seems a few things could be checked then just call `step(null, this.config.game.skipFrame)`, or `replay`, depending on the circumstances

File: engine/GameOrchestrator.js [hunk 2]
```diff
@@ -109,11 +112,50 @@ define([
...
+					this.parent.awaitCapacity(() => {
+						if(this.token !== currentToken) {
+							return;
+						}
+						this.gameActive = true;
+						this.parent.sandbox.postMessage({
+							action: 'SKIP',
+							token: this.token,
+							type: null,
+							steps: this.config.game.skipFrame,
```
Comment (davidje13) [inline engine/GameOrchestrator.js]: Is this correct? I would expect that steps should be `skipFrame - <current frame>`, not just `skipFrame`?

File: engine/GameOrchestrator.js [hunk 3]
```diff
@@ -183,8 +225,8 @@ define([
 			return this.config.game;
 		}
 
-		replay() {
-			this.begin(this.getSeed(), this.config.game.teams);
+		replay(startFrame = 0) {
+			this.begin(this.getSeed(), this.config.game.teams, startFrame);
```
Comment (davidje13) [inline engine/GameOrchestrator.js]: Not a fan of this; it seems to have caused some hacks deeper into the code too. Couldn't the usages of this be replaced with a `replay()` followed by a `step(frames)`? That would remove the need for a `startFrame` here, and would remove the need for funky hacks in the worker too.

File: engine/GameOrchestrator.js [hunk 4]
```diff
@@ -194,7 +236,7 @@ define([
 			this.begin(seed, this.config.game.teams);
 		}
 
-		begin(seed, teams) {
+		begin(seed, teams, frame = 0) {
```
Comment (davidje13) [inline engine/GameOrchestrator.js]: As noted above, I think this whole `frame` parameter when beginning a game should go

File: engine/GameOrchestrator.js [hunk 5]
```diff
@@ -346,11 +389,12 @@ define([
 				game: Object.assign({
 					seed: null,
 					teams,
+					startFrame: 0,
 				}, baseGameConfig),
 				play: Object.assign({
 					delay: 0,
 					speed: 0,
-					maxTime: 500,
+					checkbackTime: 1000,
```
Comment (davidje13) [inline engine/GameOrchestrator.js]: Why increase the default to 1000? Was this driven by anything specific?

File: engine/configuration.js [hunk 6]
```diff
@@ -24,7 +24,7 @@ define(['display/documentUtils'], (docutil) => {
 		basePlayConfig,
 		basePlayHiddenConfig: JSON.parse(meta(
 			'play-hidden-config',
-			'{"speed": -1, "maxTime": 250}'
+			'{"speed": -1, "checkbackTime": 1000}'
```
Comment (davidje13) [inline engine/configuration.js]: Again why increase the default to 1000? Was this driven by anything specific?

File: engine/gameWorker.js [hunk 7]
```diff
@@ -17,22 +17,87 @@ define(['math/Random'], (Random) => {
...
+						}
+					}
+				} catch(e) {
+					if(e === 'PAUSE') {
+						sendState(true);
+					} else {
+						throw e;
+					}
+				}
+			}
```
Comment (davidje13) [inline engine/gameWorker.js]: This whole block should be removable if my above suggestion is implemented

File: engine/gameWorker.js [hunk 8]
```diff
@@ -17,22 +17,87 @@ define(['math/Random'], (Random) => {
 			});
 		}
 
-		function begin(config) {
+		function sendIncomplete() {
+			//Call us back, we want to continue on
```
Comment (davidje13) [inline engine/gameWorker.js]: This sends partial information but carries on computing. No problem with that but it means this comment is incorrect; we must NOT call the worker back!

File: engine/gameWorker.js [hunk 9]
```diff
@@ -17,22 +17,87 @@ define(['math/Random'], (Random) => {
...
+			const now = Date.now();
+			const state = game.getState();
+			self.postMessage({
+				action: 'STEP_INCOMPLETE',
+				state: Object.assign({}, state, {
+					realWorldTime: now - time0,
+				}),
+			});
```
Comment (davidje13) [inline engine/gameWorker.js]: This code is mostly duplicated from `sendState` (and both would need to be maintained in parallel) - could they be adapted to share this code instead of duplicating it?

File: engine/gameWorker.js [hunk 10]
```diff
@@ -17,22 +17,87 @@ define(['math/Random'], (Random) => {
...
+			}
 			sendState();
 		}
 
-		function step({maxTime, steps, type}) {
-			const limit = maxTime ? (Date.now() + maxTime) : 0;
+		function step({checkbackTime, steps, type}) {
+			let limit = 0;
+			if (checkbackTime) {
+				limit = Math.floor(Date.now()/checkbackTime)*checkbackTime + checkbackTime;
```
Comment (davidje13) [inline engine/gameWorker.js]: Why are we checking back at regular intervals relative to 01/01/1970 rather than an amount of time after we were called? This seems more complicated than it needs to be.

File: engine/gameWorker.js [hunk 11]
```diff
@@ -17,22 +17,87 @@ define(['math/Random'], (Random) => {
...
+			}
 			try {
 				for(let i = 0; (steps < 0 || i < steps) && !game.isOver(); ++ i) {
 					game.step(type);
 					if(limit && Date.now() >= limit) {
-						break;
+						limit = Math.floor(Date.now()/checkbackTime)*checkbackTime + checkbackTime;
```
Comment (davidje13) [inline engine/gameWorker.js]: See above comment. Also couldn't this be replaced with `limit += checkbackTime`?

Comment (davidje13) [reply engine/gameWorker.js]: Slight risk here is that the user might try to interact with the game while we're still calculating, which previously would be allowed (old calculation would stop), but now the user must wait for the old calculation to complete before anything they do will be heard. Not sure what the best fix is, but it might be better to actually rely on the round-trip of stopping here and having the non-worker controller actually call back (as the comment on `sendIncomplete` implied)

File: engine/gameWorker.js [hunk 12]
```diff
@@ -17,22 +17,87 @@ define(['math/Random'], (Random) => {
...
+			if (checkbackTime) {
+				limit = Math.floor(Date.now()/checkbackTime)*checkbackTime + checkbackTime;
+			}
+			try {
+				while(!game.isOver() && game.getState().frame < skipFrame) {
+					game.step(type);
+					if(limit && Date.now() >= limit) {
+						let prevLim = Math.floor(Date.now()/checkbackTime)*checkbackTime;
+						limit = prevLim + checkbackTime;
+						sendIncomplete();
```
Comment (davidje13) [inline engine/gameWorker.js]: This looks duplicatey

File: engine/gameWorker.js [hunk 13]
```diff
@@ -50,13 +115,17 @@ define(['math/Random'], (Random) => {
 
 			switch(data.action) {
 			case 'BEGIN':
-				begin(data.config);
+				begin(data.config, data.checkbackTime);
```
Comment (davidje13) [inline engine/gameWorker.js]: As with earlier notes on `begin`, I think this change is not needed

File: engine/sandboxedGames.js [hunk 14]
```diff
@@ -71,11 +82,25 @@ define([
 			this.timeout = null;
 			this.waiting = true;
 			this.lastStartStep = Date.now();
+			let notForward = (type !== null && steps === null);
```
Comment (davidje13) [inline engine/sandboxedGames.js]: The condition here has changed from `type !== null || steps !== null` to `type !== null && steps === null`. This seems a strange change. I haven't looked over the logic in detail, but was this an old bug which has been fixed? a typo? something else?

File: engine/sandboxedGames.js [hunk 15]
```diff
@@ -71,11 +82,25 @@ define([
...
+		_jump(type = null, steps = null) {
+			this.timeout = null;
+			this.waiting = true;
+			this.lastStartStep = Date.now();
+			let notForward = (type !== null && steps === null);
+			this.gameWorker.postMessage({
+				action: 'SKIP',
+				type: type || '',
+				skipFrame: steps || 0,
+				checkbackTime: notForward ? 0 : this.playConfig.checkbackTime,
```
Comment (davidje13) [inline engine/sandboxedGames.js]: This looks duplicatey

File: formic.htm [hunk 16]
```diff
@@ -38,7 +38,8 @@
 	"width": 2500,
 	"height": 1000,
 	"foodRatio": 0.001,
-	"maxFrame": 30000
+	"maxFrame": 30000,
+	"skipFrame": 0
```
Comment (davidje13) [inline formic.htm]: Why is this needed in the game config?

File: games/common/components/StepperOptions.js [hunk 17]
```diff
@@ -142,7 +156,10 @@ define([
 				const maxFrame = Math.max(Math.round(this.maxFrame.value), 1);
 				this.trigger('changegame', [{maxFrame}]);
 			});
-
+			this.skipFrame.addEventListener('change', () => {
+				const skipFrame = Math.max(Math.round(this.skipFrame.value), 0);
+				this.trigger('changegame', [{skipFrame}]);
```
Comment (davidje13) [inline games/common/components/StepperOptions.js]: I don't like storing the skip frame as a game configuration state. It feels like a property of the **event**, not the **game**

File: games/common/components/StepperOptions.js [hunk 18]
```diff
@@ -216,6 +235,19 @@ define([
...
 			} else if(this.maxFrame !== docutil.document.activeElement) {
 				this.maxFrame.value = config.maxFrame;
+				if (this.skipFrame.value > this.maxFrame.value) {
+					this.skipFrame.value = this.maxFrame.value;
+				}
+			}
+			if(config.skipFrame === undefined) {
+				docutil.updateStyle(this.skipButton, {'display': 'none'});
+				docutil.updateStyle(this.skipFrame, {'display': 'none'});
+				docutil.updateAttrs(this.skipFrame, {'disabled': 'disabled'});
```
Comment (davidje13) [inline games/common/components/StepperOptions.js]: Looks like there's an idea to support skipping on a per-game basis, but I don't see why it shouldn't be supported for all games. They are all set up to be deterministic and frame-based.

File: games/formic/GameManager.js [hunk 19]
```diff
@@ -182,12 +182,14 @@ define([
 			foodRatio,
 			maxFrame,
 			teams,
+			stepFrame
```
Comment (davidje13) [inline games/formic/GameManager.js]: As noted above, I think this variable should live with the UI and be sent as part of the event, not kept on the game itself.

**Review by davidje13 — COMMENTED** [review]
There are a few messy things here and a lot of duplicated code which I think should be cleaned up, but most of the changes are good and I don't think it will take much to fix.

The headline changes I suggest:

- Remove the concept of starting a game with a specific target frame. It has led to lots of duplicate code and concepts which don't make sense for functions to know about. It even has a "hack" comment on part of it. Replace this with starting the game then immediately skipping to the desired frame. Same behaviour and much cleaner implementation.
- Target frame numbers should not be stored in the game state. They should be passed as part of the event instead (with the UI keeping track of the currently entered value if necessary)

**Comment by eaglgenes101** [comment]
One issue that I encountered when trying to separate starting a game and skipping frames was that the controller had a tendency to eat the frame skip command when the two were separated. I'll see if I can separate the two commands as suggested, but chances are fair that I might not be able to, and my self-professed hack will be refactored but remain.

**Comment by davidje13** [comment]
@eaglgenes101 the lost message issue is because you're being too polite in `GameStepper.skip`. There's no need to say `if(!this.waiting) {` there; that's just so that step events won't build up if the game is running slow. These skip events don't need that nicety. Get rid of that, and the following `Game.skip` function will work fine (tested):

```javascript
skip() {
	const frame = this.config.game.skipFrame;
	if (frame === undefined) {
		throw new Error('Skipping not enabled for this game type');
	}
	if (frame <= this.latestState.frame) {
		this.replay();
	}
	this.step(null, frame, 'SKIP');
}
```

(that also adds a 3rd parameter to `step` which can override the action name; there's probably a better way to do that though)

**Comment by davidje13** [comment]
@eaglgenes101 I've tidied up my experiment and pushed it. This change removes the hack entirely, and de-duplicates most code. It doesn't cover everything I mentioned though.

File: engine/GameOrchestrator.js (hunk 5 above)
Comment (eaglgenes101) [reply engine/GameOrchestrator.js]: Just a UI feel. The 500-ms update interval feels a bit too much like the system is running choppily, while the 1000-ms update interval gives the feel of periodic updates without excessive waiting between updates. If you want, I can change that back.

File: engine/gameWorker.js (hunk 10 above)
Comment (eaglgenes101) [reply engine/gameWorker.js]: Timing against intervals naively implies accepting whatever latency between timer expiration and response there is, letting clock drift happen. Yes, this can be accounted for with some bookkeeping, but timing against deadlines handles this issue naturally and without sweat. I personally like the synchronized updates this yields, but I get that you might disagree.

File: engine/gameWorker.js (hunk 11 above)
Comment (eaglgenes101) [reply engine/gameWorker.js]: The solution I had in mind here was a progress indicator next to a cancel button that when pushed, set a message that finished up any existing turns and then paused the game runner. I didn't get around to actually implementing that, since getting the nodejs runtime working was the more pressing matter I had.

File: engine/GameOrchestrator.js (hunk 5 above)
Comment (davidje13) [reply engine/GameOrchestrator.js]: I see. The original low-ish value is so that users can still control things when it's running slowly (i.e. pressing stop can take effect within half a second of pressing it). Increasing that to 1 second is probably fine, although it reminds me that the new `incomplete` event won't give the user that control (if they try to step at a speed of 1000 in a game which takes 0.1 seconds per frame, but get bored 10 seconds in to the 100 second task, they'll still have to wait the extra 90 seconds before they can stop it, whereas previously it would stop within a second. I'll have a think about how that issue can be avoided).

File: engine/gameWorker.js (hunk 10 above)
Comment (davidje13) [reply engine/gameWorker.js]: I think the difference in viewpoints here is because you're considering it from the perspective of a long-running task which we get updates from at regular intervals but don't need to control, whereas I'm thinking of a user who pressed the "turbo speed" button and wants the simulation to run as quickly as possible, but still leave them in control (half a second on one computer might be 1000 frames, but on another computer it might be only 10). These feel different enough in behaviour that it might be worth having 2 separate controls; one for regular update intervals (used when skipping and when running games in the background during a tournament), and the other for running as many frames as possible within a given amount of time (turbo speed).

**Comment by eaglgenes101** [comment]
I see many of your changes here will need to have corresponding changes to the nodejs runtime. What else needs to be done here?

**Comment by davidje13** [comment]
@eaglgenes101 I was afraid of that. I think it's important to keep the `maxDuration` / `checkbackInterval` split for the web player where user interaction is the main concern, but in the node runtime `maxDuration` could be dropped (which might make this just a simple name change from `checkbackTime` to `checkbackInterval`?)

I changed a couple of APIs to take named parameters to reduce chances of error, but I don't think there's anything too breaking.

As for what else to do, I'm going to take a final look at this now that I've had some time away from the code to do a final review (and if there are any issues you've spotted please let me know). But functionally I'm happy to merge it now.

**Comment by davidje13** [comment]
From a quick look-through I'm happy with these changes.

I'm conscious that my latest changes touched most of this, so I'll leave this open for @eaglgenes101 or @dzaima to review and make sure it still accomplishes their original goals and they're happy with the implementation. In the meantime I'll rebase the nodejs changes and continue breaking up that branch.

**Comment by eaglgenes101** [comment]
At least on the browser side I think we can give this the go-ahead. I have some differences in preferences, but it's a pretty good realization of the feature as intended.
